linera-core/src/client/validator_manager/in_flight_tracker.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, fmt::Debug, sync::Arc};
+
+use linera_base::time::{Duration, Instant};
+use tokio::sync::broadcast;
+
+use super::{
+    cache::SubsumingKey,
+    request::{RequestKey, RequestResult},
+};
+use crate::node::NodeError;
+
+/// Tracks in-flight requests to deduplicate concurrent requests for the same data.
+///
+/// This structure manages a map of request keys to in-flight entries, each containing
+/// broadcast senders for notifying waiters when a request completes, as well as timing
+/// information and alternative data sources.
+#[derive(Debug, Clone)]
+pub(super) struct InFlightTracker<N> {
+    /// Maps request keys to in-flight entries containing broadcast senders and metadata
+    entries: Arc<tokio::sync::RwLock<HashMap<RequestKey, InFlightEntry<N>>>>,
+    /// Maximum duration before an in-flight request is considered stale and deduplication is skipped
+    timeout: Duration,
+}
+
+impl<N: Clone> InFlightTracker<N> {
+    /// Creates a new `InFlightTracker` with the specified timeout.
+    ///
+    /// # Arguments
+    /// - `timeout`: Maximum duration before an in-flight request is considered too old to deduplicate against
+    pub(super) fn new(timeout: Duration) -> Self {
+        Self {
+            entries: Arc::new(tokio::sync::RwLock::new(HashMap::new())),
+            timeout,
+        }
+    }
+
+    /// Attempts to subscribe to an existing in-flight request (exact or subsuming match).
+    ///
+    /// Searches for either an exact key match or a subsuming request (whose result would
+    /// contain all the data needed by this request). Returns information about which type
+    /// of match was found, along with subscription details.
+    ///
+    /// # Arguments
+    /// - `key`: The request key to look up
+    ///
+    /// # Returns
+    /// - `None`: No matching in-flight request found. Also returned if the found request is stale (exceeds timeout).
+    /// - `Some(InFlightMatch::Subsuming { key, outcome })`: Subsuming request found
+    pub(super) async fn try_subscribe(&self, key: &RequestKey) -> Option<InFlightMatch> {
+        let in_flight = self.entries.read().await;
+
+        if let Some(entry) = in_flight.get(key) {
+            let elapsed = Instant::now().duration_since(entry.started_at);
+
+            if elapsed <= self.timeout {
+                return Some(InFlightMatch::Exact(Subscribed(entry.sender.subscribe())));
+            }
+        }
+
+        // Sometimes a request key may not have the exact match but may be subsumed by a larger one.
+        for (in_flight_key, entry) in in_flight.iter() {
```

Comment:
Not sure how many in-flight requests we are expecting, but if we had the map by chain ID and/or type of request, we wouldn't have to iterate over all of them.

Reply:
I think @deuszx measured that this PR already improves performance a lot on the testnet, so it seems like the CPU overhead is small compared to the bandwidth gains?